### PR TITLE
backend/azurerm: support for authenticating via SAS Tokens

### PR DESCRIPTION
**Pull request by tombuildsstuff**
This PR adds support for authenticating using a SAS Token, which allows limited access permissions to be used to access a Storage Account.

Needs to be merged after #19433

### CODE REVIEW

File: backend/remote-state/azure/backend_test.go [hunk 1]
```diff
@@ -54,6 +54,35 @@ func TestBackendAccessKeyBasic(t *testing.T) {
...
+}
+
+func TestBackendManagedServiceIdentityBasic(t *testing.T) {
+	testAccAzureBackendRunningInAzure(t)
+	rs := acctest.RandString(4)
+	res := testResourceNames(rs, "testState")
+	armClient := buildTestClient(t, res)
+
+	ctx := context.TODO()
+	err := armClient.buildTestResources(ctx, &res)
```
Comment (alexsomesan) [inline backend/remote-state/azure/backend_test.go]: Merge with checking for `nil`?

File: backend/remote-state/azure/backend_test.go [hunk 2]
```diff
@@ -54,6 +54,35 @@ func TestBackendAccessKeyBasic(t *testing.T) {
...
+	if err != nil {
+		armClient.destroyTestResources(ctx, res)
+		t.Fatalf("Error creating Test Resources: %q", err)
```
Comment (alexsomesan) [inline backend/remote-state/azure/backend_test.go]: `Fatalf()` (actually FailNow() inside it) will unroll the defer stack and execute it before exiting.
This means you could just put the `defer` before checking err and remove the explicit call to `armClient.destroyTestResources` above.

File: backend/remote-state/azure/backend_test.go [hunk 3]
```diff
@@ -87,6 +116,36 @@ func TestBackendManagedServiceIdentityBasic(t *testing.T) {
...
 
+func TestBackendSASTokenBasic(t *testing.T) {
+	testAccAzureBackend(t)
+	rs := acctest.RandString(4)
+	res := testResourceNames(rs, "testState")
+	armClient := buildTestClient(t, res)
+
+	ctx := context.TODO()
+	err := armClient.buildTestResources(ctx, &res)
+	if err != nil {
```
Comment (alexsomesan) [inline backend/remote-state/azure/backend_test.go]: Same observation about `defer` as before.

File: backend/remote-state/azure/client_test.go [hunk 4]
```diff
@@ -35,6 +35,40 @@ func TestRemoteClientAccessKeyBasic(t *testing.T) {
...
+
+func TestRemoteClientManagedServiceIdentityBasic(t *testing.T) {
+	testAccAzureBackendRunningInAzure(t)
+	rs := acctest.RandString(4)
+	res := testResourceNames(rs, "testState")
+	armClient := buildTestClient(t, res)
+
+	ctx := context.TODO()
+	err := armClient.buildTestResources(ctx, &res)
+	if err != nil {
```
Comment (alexsomesan) [inline backend/remote-state/azure/client_test.go]: The `defer` stuff again.

File: backend/remote-state/azure/client_test.go [hunk 5]
```diff
@@ -78,6 +112,41 @@ func TestRemoteClientManagedServiceIdentityBasic(t *testing.T) {
...
 
+func TestRemoteClientSasTokenBasic(t *testing.T) {
+	testAccAzureBackend(t)
+	rs := acctest.RandString(4)
+	res := testResourceNames(rs, "testState")
+	armClient := buildTestClient(t, res)
+
+	ctx := context.TODO()
+	err := armClient.buildTestResources(ctx, &res)
+	if err != nil {
```
Comment (alexsomesan) [inline backend/remote-state/azure/client_test.go]: also `defer`

**Comment by tombuildsstuff** [comment]
Ignoring the known failing tests (which are fixed by #19441) - the tests otherwise look good:

screenshot 2018-11-22 at 17 43 30

**Comment by ps23** [comment]
Is possible to get a working example for that in .tf style?

```bash
terraform {
  backend "azurerm" {
    sas_token            = "xxx"
  }
}
```

Does it aim at the SAS token for the container (`az storage container generate-sas ...`) or the blob (`az storage blob generate-sas ...`) itself?

**Comment by abelal83** [comment]
I see this PR was merged before tf v0.11.11 was built but sas_token doesn't seem to be supported?

**Comment by tombuildsstuff** [comment]
@abelal83 unfortunately this PR is not included in v0.11.11 - it will ship as a part of Terraform 0.12

**Comment by ps23** [comment]
> @abelal83 unfortunately this PR is not included in v0.11.11 - it will ship as a part of Terraform 0.12

Is it in the current alpha?

**Comment by abelal83** [comment]
That's great to hear @tombuildsstuff

**Comment by abelal83** [comment]
@ps23 don't think it is. I had a look at the build artifacts and sas key support didn't make it in.

**Comment by tombuildsstuff** [comment]
@ps23 it appears this shipped in Terraform 0.12 Alpha 3 (and later): https://github.com/hashicorp/terraform/blob/master/CHANGELOG.md#0120-alpha3-december-6-2018

**Pull request merged**